ports/javascript/Makefile

```diff
@@ -13,13 +13,29 @@ INC += -I.
 INC += -I$(TOP)
 INC += -I$(BUILD)
 
-CPP = clang -E
-
 ifdef EMSCRIPTEN
-	CPP += -D__EMSCRIPTEN__
-	CPP += --sysroot $(EMSCRIPTEN)/system
-	CPP += -isystem $(EMSCRIPTEN)/system/include/libc
-	CPP += -cxx-isystem $(EMSCRIPTEN)/system/include/libcxx
+	# only for profiling, remove -s EMTERPRETIFY_ADVISE=1 when your EMTERPRETIFY_WHITELIST is ok
+	# not using an emterpreting list *is* bad and will lead to poor performance and huge binary.
+	ifdef ASYNC
+		CFLAGS += -D__EMTERPRETER__=1 
+		CFLAGS += -s EMTERPRETIFY=1 -s EMTERPRETIFY_ASYNC=1 -s 'EMTERPRETIFY_FILE="micropython.binary"'
+		CFLAGS += -s EMTERPRETIFY_SYNCLIST='["_mp_execute_bytecode"]' -s EMTERPRETIFY_ADVISE=1
+	endif
+
+	CC = emcc
+	CPP = clang -E -D__CPP__ -D__EMSCRIPTEN__
+	CPP += --sysroot $(EMSCRIPTEN)/system	
+	CPP += $(addprefix -isystem, $(shell env LC_ALL=C $(CC) $(CFLAGS_EXTRA) -E -x c++ /dev/null -v 2>&1 |sed -e '/^\#include <...>/,/^End of search/{ //!b };d'))
+	# Act like 'emcc'
+	CPP += -U__i386 -U__i386 -Ui386 -U__SSE -U__SSE_MATH -U__SSE2 -U__SSE2_MATH -U__MMX__ -U__SSE__ -U__SSE_MATH__ -U__SSE2__ -U__SSE2_MATH__
```

Comment:
The manual undefining of macros should be replaced by the trick I described here: https://github.com/micropython/micropython/pull/4875#pullrequestreview-259447603

Reply:
@embeddedt maybe push the all the tricks at once first here https://github.com/littlevgl/lv_micropython/blob/pmpp_test/ports/javascript/Makefile so i can try them with child modules ?

Reply:
They were already there: https://github.com/littlevgl/lv_micropython/commit/3bd1eabc7fd3551ce8fe35dfc12b492d5ce1f998

Reply:
iirc i tried those but i had a problem with macro redef, but adding -Wno-macro-redefined to $(CPP) seems to fix it

Reply:
Disabling that warning is not a good solution to the problem (it just hides it). What macros are you getting the warning on?

Reply:
```
Use make V=1 or set BUILD_VERBOSE in your environment to increase build verbosity.
Including User C Module from cmod/example
Makefile:178: warning: overriding recipe for target 'clean'
../../py/mkrules.mk:157: warning: ignoring old recipe for target 'clean'
Makefile:300: warning: overriding recipe for target 'micropython.html'
../../py/mkrules.mk:128: warning: ignoring old recipe for target 'micropython.html'
mkdir -p build/genhdr
GEN build/genhdr/mpversion.h
GEN build/genhdr/moduledefs.h
GEN build/genhdr/qstr.i.last
build/clang_predefs.h:24:9: error: '__CLANG_ATOMIC_LLONG_LOCK_FREE' macro redefined [-Werror,-Wmacro-redefined]
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 1
        ^
<built-in>:300:9: note: previous definition is here
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 2
        ^
build/clang_predefs.h:85:9: error: '__GCC_ATOMIC_LLONG_LOCK_FREE' macro redefined [-Werror,-Wmacro-redefined]
#define __GCC_ATOMIC_LLONG_LOCK_FREE 1
        ^
<built-in>:310:9: note: previous definition is here
#define __GCC_ATOMIC_LLONG_LOCK_FREE 2
        ^
build/clang_predefs.h:107:9: error: '__INT64_C_SUFFIX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_C_SUFFIX__ LL
        ^
<built-in>:175:9: note: previous definition is here
#define __INT64_C_SUFFIX__ L
        ^
build/clang_predefs.h:108:9: error: '__INT64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_FMTd__ "lld"
        ^
<built-in>:173:9: note: previous definition is here
#define __INT64_FMTd__ "ld"
        ^
build/clang_predefs.h:109:9: error: '__INT64_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_FMTi__ "lli"
        ^
<built-in>:174:9: note: previous definition is here
#define __INT64_FMTi__ "li"
        ^
build/clang_predefs.h:110:9: error: '__INT64_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_MAX__ 9223372036854775807LL
        ^
<built-in>:207:9: note: previous definition is here
#define __INT64_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:111:9: error: '__INT64_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT64_TYPE__ long long int
        ^
<built-in>:172:9: note: previous definition is here
#define __INT64_TYPE__ long int
        ^
build/clang_predefs.h:117:9: error: '__INTMAX_C_SUFFIX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_C_SUFFIX__ LL
        ^
<built-in>:64:9: note: previous definition is here
#define __INTMAX_C_SUFFIX__ L
        ^
build/clang_predefs.h:118:9: error: '__INTMAX_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_FMTd__ "lld"
        ^
<built-in>:62:9: note: previous definition is here
#define __INTMAX_FMTd__ "ld"
        ^
build/clang_predefs.h:119:9: error: '__INTMAX_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_FMTi__ "lli"
        ^
<built-in>:63:9: note: previous definition is here
#define __INTMAX_FMTi__ "li"
        ^
build/clang_predefs.h:120:9: error: '__INTMAX_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_MAX__ 9223372036854775807LL
        ^
<built-in>:42:9: note: previous definition is here
#define __INTMAX_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:121:9: error: '__INTMAX_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTMAX_TYPE__ long long int
        ^
<built-in>:61:9: note: previous definition is here
#define __INTMAX_TYPE__ long int
        ^
build/clang_predefs.h:125:9: error: '__INTPTR_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTPTR_MAX__ 2147483647L
        ^
<built-in>:46:9: note: previous definition is here
#define __INTPTR_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:127:9: error: '__INTPTR_WIDTH__' macro redefined [-Werror,-Wmacro-redefined]
#define __INTPTR_WIDTH__ 32
        ^
<built-in>:79:9: note: previous definition is here
#define __INTPTR_WIDTH__ 64
        ^
build/clang_predefs.h:136:9: error: '__INT_FAST64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_FMTd__ "lld"
        ^
<built-in>:280:9: note: previous definition is here
#define __INT_FAST64_FMTd__ "ld"
        ^
build/clang_predefs.h:137:9: error: '__INT_FAST64_FMTi__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_FMTi__ "lli"
        ^
<built-in>:281:9: note: previous definition is here
#define __INT_FAST64_FMTi__ "li"
        ^
build/clang_predefs.h:138:9: error: '__INT_FAST64_MAX__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_MAX__ 9223372036854775807LL
        ^
<built-in>:279:9: note: previous definition is here
#define __INT_FAST64_MAX__ 9223372036854775807L
        ^
build/clang_predefs.h:139:9: error: '__INT_FAST64_TYPE__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_FAST64_TYPE__ long long int
        ^
<built-in>:278:9: note: previous definition is here
#define __INT_FAST64_TYPE__ long int
        ^
build/clang_predefs.h:152:9: error: '__INT_LEAST64_FMTd__' macro redefined [-Werror,-Wmacro-redefined]
#define __INT_LEAST64_FMTd__ "lld"
        ^
<built-in>:240:9: note: previous definition is here
#define __INT_LEAST64_FMTd__ "ld"
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
../../py/mkrules.mk:74: recipe for target 'build/genhdr/qstr.i.last' failed
make: *** [build/genhdr/qstr.i.last] Error 1
make: *** Deleting file 'build/genhdr/qstr.i.last'
```

Reply:
CPP has to be `clang -E -undef`.

Reply:
indeed, seems better

Reply:
hmm on my port i get that with -undef , while -Wno-macro-redefined pass the build
```CC ../../extmod/moductypes.c
../../extmod/moductypes.c:747:19: error: use of undeclared identifier 'MP_QSTR_SHORT'
    { MP_ROM_QSTR(MP_QSTR_SHORT), MP_ROM_INT(TYPE2SMALLINT(INT16, 4)) },
                  ^
../../extmod/moductypes.c:748:19: error: use of undeclared identifier 'MP_QSTR_USHORT'
    { MP_ROM_QSTR(MP_QSTR_USHORT), MP_ROM_INT(TYPE2SMALLINT(UINT16, 4)) },
                  ^
../../extmod/moductypes.c:751:19: error: use of undeclared identifier 'MP_QSTR_INT'; did you mean 'MP_QSTR_INT8'?
    { MP_ROM_QSTR(MP_QSTR_INT), MP_ROM_INT(TYPE2SMALLINT(INT32, 4)) },
                  ^~~~~~~~~~~
                  MP_QSTR_INT8
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:297:6: note: 'MP_QSTR_INT8' declared here
QDEF(MP_QSTR_INT8, (const byte*)"\xce\x04" "INT8")
     ^
../../extmod/moductypes.c:752:19: error: use of undeclared identifier 'MP_QSTR_UINT'; did you mean 'MP_QSTR_UINT8'?
    { MP_ROM_QSTR(MP_QSTR_UINT), MP_ROM_INT(TYPE2SMALLINT(UINT32, 4)) },
                  ^~~~~~~~~~~~
                  MP_QSTR_UINT8
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:309:6: note: 'MP_QSTR_UINT8' declared here
QDEF(MP_QSTR_UINT8, (const byte*)"\xbb\x05" "UINT8")
     ^
../../extmod/moductypes.c:755:19: error: use of undeclared identifier 'MP_QSTR_LONG'; did you mean 'MP_QSTR_VOID'?
    { MP_ROM_QSTR(MP_QSTR_LONG), MP_ROM_INT(TYPE2SMALLINT(INT32, 4)) },
                  ^~~~~~~~~~~~
                  MP_QSTR_VOID
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:310:6: note: 'MP_QSTR_VOID' declared here
QDEF(MP_QSTR_VOID, (const byte*)"\x31\x04" "VOID")
     ^
../../extmod/moductypes.c:756:19: error: use of undeclared identifier 'MP_QSTR_ULONG'
    { MP_ROM_QSTR(MP_QSTR_ULONG), MP_ROM_INT(TYPE2SMALLINT(UINT32, 4)) },
                  ^
../../extmod/moductypes.c:762:19: error: use of undeclared identifier 'MP_QSTR_LONGLONG'
    { MP_ROM_QSTR(MP_QSTR_LONGLONG), MP_ROM_INT(TYPE2SMALLINT(INT64, 4)) },
                  ^
../../extmod/moductypes.c:763:19: error: use of undeclared identifier 'MP_QSTR_ULONGLONG'; did you mean
      'MP_QSTR_ENOTCONN'?
    { MP_ROM_QSTR(MP_QSTR_ULONGLONG), MP_ROM_INT(TYPE2SMALLINT(UINT64, 4)) },
                  ^~~~~~~~~~~~~~~~~
                  MP_QSTR_ENOTCONN
../../py/obj.h:248:40: note: expanded from macro 'MP_ROM_QSTR'
#define MP_ROM_QSTR(q) MP_OBJ_NEW_QSTR(q)
                                       ^
../../py/obj.h:92:56: note: expanded from macro 'MP_OBJ_NEW_QSTR'
#define MP_OBJ_NEW_QSTR(qst) ((mp_obj_t)((((mp_uint_t)(qst)) << 2) | 2))
                                                       ^
build/genhdr/qstrdefs.generated.h:287:6: note: 'MP_QSTR_ENOTCONN' declared here
QDEF(MP_QSTR_ENOTCONN, (const byte*)"\x79\x08" "ENOTCONN")
     ^
../../extmod/moductypes.c:771:8: error: invalid application of 'sizeof' to an incomplete type
      'const mp_rom_map_elem_t []'
STATIC MP_DEFINE_CONST_DICT(mp_module_uctypes_globals, mp_module_uctypes_globals_table);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:314:21: note: expanded from macro 'MP_DEFINE_CONST_DICT'
            .used = MP_ARRAY_SIZE(table_name), \
                    ^~~~~~~~~~~~~~~~~~~~~~~~~
../../py/misc.h:111:33: note: expanded from macro 'MP_ARRAY_SIZE'
#define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
                                ^~~
../../extmod/moductypes.c:771:8: error: invalid application of 'sizeof' to an incomplete type
      'const mp_rom_map_elem_t []'
STATIC MP_DEFINE_CONST_DICT(mp_module_uctypes_globals, mp_module_uctypes_globals_table);
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../py/obj.h:315:22: note: expanded from macro 'MP_DEFINE_CONST_DICT'
            .alloc = MP_ARRAY_SIZE(table_name), \
                     ^~~~~~~~~~~~~~~~~~~~~~~~~
../../py/misc.h:111:33: note: expanded from macro 'MP_ARRAY_SIZE'
#define MP_ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
                                ^~~
10 errors generated.
shared:ERROR: compiler frontend failed to generate LLVM bitcode, halting
../../py/mkrules.mk:47: recipe for target 'build/extmod/moductypes.o' failed
make: *** [build/extmod/moductypes.o] Error 1
```

Reply:
`-Wno-macro-redefined` is just hiding the issue because we don't want any predefined macros coming from `clang`.

Can you clean and rebuild?

Reply:
my clang was missing some macro cleared by -undef, but ```-include $(BUILD)/clang_predefs.h``` can re add them.